qiskit/utils/deprecation.py

```diff
@@ -89,27 +102,96 @@ def wrapper(*args, **kwargs):
 def _rename_kwargs(
     func_name: str,
     kwargs: Dict[str, Any],
-    kwarg_map: Dict[str, str],
+    old_kwarg_to_msg: Dict[str, str],
+    kwarg_map: Dict[str, Optional[str]],
     category: Type[Warning] = DeprecationWarning,
 ) -> None:
     for old_arg, new_arg in kwarg_map.items():
-        if old_arg in kwargs:
-            if new_arg in kwargs:
-                raise TypeError(f"{func_name} received both {new_arg} and {old_arg} (deprecated).")
-
-            if new_arg is None:
-                warnings.warn(
-                    f"{func_name} keyword argument {old_arg} is deprecated and "
-                    "will in future be removed.",
-                    category=category,
-                    stacklevel=3,
-                )
-            else:
-                warnings.warn(
-                    f"{func_name} keyword argument {old_arg} is deprecated and "
-                    f"replaced with {new_arg}.",
-                    category=category,
-                    stacklevel=3,
-                )
-
-                kwargs[new_arg] = kwargs.pop(old_arg)
+        if old_arg not in kwargs:
+            continue
+        if new_arg in kwargs:
+            raise TypeError(f"{func_name} received both {new_arg} and {old_arg} (deprecated).")
+        warnings.warn(old_kwarg_to_msg[old_arg], category=category, stacklevel=3)
+        if new_arg is not None:
+            kwargs[new_arg] = kwargs.pop(old_arg)
+
+
+def _add_deprecation_to_docstring(
+    func: Callable, msg: str, *, since: Optional[str], pending: bool
+) -> None:
+    if "\n" in msg:
+        raise ValueError(
+            "Deprecation messages cannot contain new lines (`\\n`), but the deprecation for "
+            f'{func.__qualname__} had them. Usually this happens when using `"""` multiline '
+            f"strings; instead, use string concatenation.\n\n"
+            "This is a simplification to facilitate deprecation messages being added to our "
+            "documentation. If you have a compelling reason to need "
+            "new lines, feel free to improve this function or open a request at "
+            "https://github.com/Qiskit/qiskit-terra/issues."
+        )
+
+    if since is None:
+        version_str = "unknown"
+    else:
+        version_str = f"{since}_pending" if pending else since
```

Comment:
How do you deal with the pending deprecations? Is this something sphinx supports out of the box?

Reply:
Sphinx doesn't have any builtin support for pending deprecations. So, the idea I came up with is to add `_pending` to the end of the since version string to indicate the deprecation hasn't yet happened. The deprecation message itself will also make this clear, like saying "is pending deprecation"